File: legacy/flutter/lib/domain/rules/budget_calculator.dart
/// Teto do dia e orçamento da janela. Dado puro, sem widget e sem rede.
///
/// Crédito de treino não tem cap de kcal. Treino ausente (`null`) é diferente
/// de treino informado como 0: os dois creditam 0, mas só o null significa
/// que o dia ainda não foi preenchido.
library;

/// Como o perfil guarda o teto. Cada variante carrega só os números que usa.
sealed class TetoPerfil {
  const TetoPerfil();

  /// Teto base da [data], antes do crédito de treino.
  int tetoNaData(DateTime data);
}

/// O mesmo teto em qualquer dia da semana.
final class TetoMesmoTodosOsDias extends TetoPerfil {
  const TetoMesmoTodosOsDias(this.kcal);

  final int kcal;

  @override
  int tetoNaData(DateTime data) => kcal;
}

/// Segunda–sexta contra sábado–domingo. [DateTime.weekday] 1–5 vs 6–7.
final class TetoUtilFds extends TetoPerfil {
  const TetoUtilFds({required this.util, required this.fds});

  final int util;
  final int fds;

  @override
  int tetoNaData(DateTime data) {
    final fimDeSemana = data.weekday >= DateTime.saturday;
    return fimDeSemana ? fds : util;
  }
}

/// Um teto por dia, de segunda a domingo. Não colapsa a semana num único slot.
final class TetoSeteDias extends TetoPerfil {
  const TetoSeteDias({
    required this.segunda,
    required this.terca,
    required this.quarta,
    required this.quinta,
    required this.sexta,
    required this.sabado,
    required this.domingo,
  });

  final int segunda;
  final int terca;
  final int quarta;
  final int quinta;
  final int sexta;
  final int sabado;
  final int domingo;

  @override
  int tetoNaData(DateTime data) {
    return switch (data.weekday) {
      DateTime.monday => segunda,
      DateTime.tuesday => terca,
      DateTime.wednesday => quarta,
      DateTime.thursday => quinta,
      DateTime.friday => sexta,
      DateTime.saturday => sabado,
      DateTime.sunday => domingo,
      _ => throw ArgumentError.value(
        data.weekday,
        'data',
        'dia da semana fora de 1..7',
      ),
    };
  }
}

/// Política de eat-back do treino. Parcial usa o percentual informado, sem clamp.
enum PoliticaCreditoTreino { zero, parcial, cem }

/// Entrada de um dia. [reservaProximas] já vem pronta — esta camada não soma refeições futuras.
final class EntradaOrcamento {
  const EntradaOrcamento({
    required this.data,
    required this.perfil,
    required this.politica,
    this.percentual,
    this.treinoKcal,
    required this.consumido,
    required this.reservaProximas,
  });

  final DateTime data;
  final TetoPerfil perfil;
  final PoliticaCreditoTreino politica;

  /// Percentual digitado (50 significa 50%). Só entra quando [politica] é parcial.
  final num? percentual;

  /// Kcal do treino do dia. Null = não informado.
  final num? treinoKcal;

  final num consumido;
  final num reservaProximas;
}

/// Resultado de um único cálculo: base, crédito, teto efetivo e orçamento da janela.
final class ResultadoOrcamento {
  const ResultadoOrcamento({
    required this.tetoBase,
    required this.creditoTreino,
    required this.tetoEfetivo,
    required this.orcamentoJanela,
  });

  final int tetoBase;
  final num creditoTreino;
  final num tetoEfetivo;
  final num orcamentoJanela;

  @override
  String toString() {
    return 'ResultadoOrcamento(tetoBase: $tetoBase, '
        'creditoTreino: $creditoTreino, tetoEfetivo: $tetoEfetivo, '
        'orcamentoJanela: $orcamentoJanela)';
  }
}

/// Calculadora sem estado. Cada chamada parte só da [entrada].
final class BudgetCalculator {
  const BudgetCalculator();

  ResultadoOrcamento calcular(EntradaOrcamento entrada) {
    final tetoBase = entrada.perfil.tetoNaData(entrada.data);
    final creditoTreino = _creditoTreino(
      politica: entrada.politica,
      percentual: entrada.percentual,
      treinoKcal: entrada.treinoKcal,
    );
    final tetoEfetivo = tetoBase + creditoTreino;
    final bruto = tetoEfetivo - entrada.consumido - entrada.reservaProximas;
    return ResultadoOrcamento(
      tetoBase: tetoBase,
      creditoTreino: creditoTreino,
      tetoEfetivo: tetoEfetivo,
      orcamentoJanela: bruto < 0 ? 0 : bruto,
    );
  }
}

/// Política 0, ou treino do dia ainda não informado, zera o crédito.
/// Parcial é o produto cru `treino * (pct / 100)`. 100% devolve o treino inteiro.
num _creditoTreino({
  required PoliticaCreditoTreino politica,
  required num? percentual,
  required num? treinoKcal,
}) {
  if (politica == PoliticaCreditoTreino.zero || treinoKcal == null) {
    return 0;
  }
  if (politica == PoliticaCreditoTreino.cem) {
    return treinoKcal;
  }
  if (percentual == null) {
    throw ArgumentError.notNull('percentual');
  }
  return treinoKcal * percentual / 100;
}

File: apps/rn/src/domain/budgetCalculator.ts
/** Teto do dia e orçamento da janela. Dado puro, sem React e sem rede. */

export const ZONA = "America/Sao_Paulo";
const OFFSET_MS = 3 * 60 * 60 * 1000;

export type DataCivil = { y: number; m: number; d: number };

export type TetoPerfil =
  | { tipo: "mesmo"; kcal: number }
  | { tipo: "util"; util: number; fds: number }
  | {
      tipo: "sete";
      segunda: number;
      terca: number;
      quarta: number;
      quinta: number;
      sexta: number;
      sabado: number;
      domingo: number;
    };

export type PoliticaCredito = "zero" | "parcial" | "cem";

export type EntradaOrcamento = {
  data: DataCivil;
  perfil: TetoPerfil;
  politica: PoliticaCredito;
  percentual?: number | null;
  treinoKcal?: number | null;
  consumido?: number;
  reservaProximas?: number;
};

export type ResultadoOrcamento = {
  tetoBase: number;
  creditoTreino: number;
  tetoEfetivo: number;
  orcamentoJanela: number;
};

/** 1 = segunda … 7 = domingo. Meio-dia UTC evita a borda do fuso da máquina. */
export function weekday(data: DataCivil): number {
  const js = new Date(Date.UTC(data.y, data.m - 1, data.d, 12)).getUTCDay();
  return js === 0 ? 7 : js;
}

function deslocar(instant: Date): Date {
  return new Date(instant.getTime() - OFFSET_MS);
}

/** Instant UTC → data civil em America/Sao_Paulo (UTC−3, sem horário de verão). */
export function dataSaoPaulo(instantIso: string): DataCivil {
  const deslocado = deslocar(new Date(Date.parse(instantIso)));
  return {
    y: deslocado.getUTCFullYear(),
    m: deslocado.getUTCMonth() + 1,
    d: deslocado.getUTCDate(),
  };
}

export function horaSaoPaulo(instant: Date = new Date()): number {
  return deslocar(instant).getUTCHours();
}

/** ISO local com offset −03:00, para o POST de estimate/fit. */
export function localTimeSaoPaulo(instant: Date = new Date()): string {
  const sp = deslocar(instant);
  const p = (n: number) => String(n).padStart(2, "0");
  return `${sp.getUTCFullYear()}-${p(sp.getUTCMonth() + 1)}-${p(sp.getUTCDate())}T${p(sp.getUTCHours())}:${p(sp.getUTCMinutes())}:${p(sp.getUTCSeconds())}-03:00`;
}

export function tetoNaData(perfil: TetoPerfil, data: DataCivil): number {
  if (perfil.tipo === "mesmo") return perfil.kcal;
  const dia = weekday(data);
  if (perfil.tipo === "util") return dia >= 6 ? perfil.fds : perfil.util;
  switch (dia) {
    case 1:
      return perfil.segunda;
    case 2:
      return perfil.terca;
    case 3:
      return perfil.quarta;
    case 4:
      return perfil.quinta;
    case 5:
      return perfil.sexta;
    case 6:
      return perfil.sabado;
    default:
      return perfil.domingo;
  }
}

/** Política 0, ou treino ainda não informado, zera o crédito. Sem cap. */
export function creditoTreino(
  politica: PoliticaCredito,
  percentual: number | null | undefined,
  treinoKcal: number | null | undefined,
): number {
  if (politica === "zero" || treinoKcal == null) return 0;
  if (politica === "cem") return treinoKcal;
  if (percentual == null) throw new Error("percentual");
  return (treinoKcal * percentual) / 100;
}

export function calcular(entrada: EntradaOrcamento): ResultadoOrcamento {
  const tetoBase = tetoNaData(entrada.perfil, entrada.data);
  const credito = creditoTreino(entrada.politica, entrada.percentual, entrada.treinoKcal);
  const tetoEfetivo = tetoBase + credito;
  const bruto = tetoEfetivo - (entrada.consumido ?? 0) - (entrada.reservaProximas ?? 0);
  return {
    tetoBase,
    creditoTreino: credito,
    tetoEfetivo,
    orcamentoJanela: bruto < 0 ? 0 : bruto,
  };
}
